### PR TITLE
Add user help on TLS 1.0/1.1 deprecation

### PR DESCRIPTION
**Pull request by brainwane**
Add context and instructions for users affected by the TLS 1.0/1.1 deprecation.

Followup to #3411, #3293.

### CODE REVIEW

**Comment by brainwane** [comment]
@ned-deily I would appreciate a review.

@mitsuhiko Thanks for spurring this.

**Comment by brainwane** [comment]
@ampersamd would this help text have helped you?

**Comment by brainwane** [comment]
@jonparrott is there part of this that should actually live on PyPUG?

**Comment by ampersamd** [comment]
@brainwane That would've definitely helped

**Comment by jonparrott** [comment]
> @jonparrott is there part of this that should actually live on PyPUG?

I'm not sure where it would live. I could see this being helpful in the pip docs, however.

**Comment by ned-deily** [comment]
Sigh!  This is complicated for macOS, or at least it seems so to me.  Perhaps I'm making it too complicated.  But, it seems to me you want the instructions to be close to bullet-proof as the users most likely to be consulting them are those who will likely have the least experience trying to wade through this kind of stuff.

The first issue is that you don't know for sure what typing `python` is going to do.  On a vanilla macOS system (if the user hasn't installed anything yet), it will likely invoke an Apple-supplied system Python; on recent macOS releases, that would be a version of Python2.7.  Up through the current macOS (10.13) Apple has never shipped `pip` so you can't selfupgrade it.  The simple instructions in the above email:
`curl https://bootstrap.pypa.io/get-pip.py | python`
appear to work without giving an error but actually don't work (at least on the systems I tried):

```
$ curl https://bootstrap.pypa.io/get-pip.py | python
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100 1602k  100 1602k    0     0  5723k      0 --:--:-- --:--:-- --:--:-- 5744k
Collecting pip
  Downloading https://files.pythonhosted.org/packages/62/a1/0d452b6901b0157a0134fd27ba89bf95a857fbda64ba52e1ca2cf61d8412/pip-10.0.0-py2.py3-none-any.whl (1.3MB)
    100% |████████████████████████████████| 1.3MB 4.1MB/s
Collecting wheel
  Downloading https://files.pythonhosted.org/packages/1b/d2/22cde5ea9af055f81814f9f2545f5ed8a053eb749c08d186b369959189a8/wheel-0.31.0-py2.py3-none-any.whl (41kB)
    100% |████████████████████████████████| 51kB 12.9MB/s
matplotlib 1.3.1 requires nose, which is not installed.
matplotlib 1.3.1 requires tornado, which is not installed.
Installing collected packages: pip, wheel

$ pip
-bash: pip: command not found
```

It appears that you need to use administrator privileges to install for the system Python.  So something like this should (note the sudo :( ):
```
$ curl https://bootstrap.pypa.io/get-pip.py | sudo python
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0Password:
100 1602k  100 1602k    0     0   138k      0  0:00:11  0:00:11 --:--:--  138k
The directory '/Users/sysadmin/Library/Caches/pip/http' or its parent directory is not owned by the current user and the cache has been disabled. Please check the permissions and owner of that directory. If executing pip with sudo, you may want sudo's -H flag.
The directory '/Users/sysadmin/Library/Caches/pip' or its parent directory is not owned by the current user and caching wheels has been disabled. check the permissions and owner of that directory. If executing pip with sudo, you may want sudo's -H flag.
Collecting pip
  Downloading https://files.pythonhosted.org/packages/62/a1/0d452b6901b0157a0134fd27ba89bf95a857fbda64ba52e1ca2cf61d8412/pip-10.0.0-py2.py3-none-any.whl (1.3MB)
    100% |████████████████████████████████| 1.3MB 4.9MB/s
Collecting wheel
  Downloading https://files.pythonhosted.org/packages/1b/d2/22cde5ea9af055f81814f9f2545f5ed8a053eb749c08d186b369959189a8/wheel-0.31.0-py2.py3-none-any.whl (41kB)
    100% |████████████████████████████████| 51kB 17.0MB/s
matplotlib 1.3.1 requires nose, which is not installed.
matplotlib 1.3.1 requires tornado, which is not installed.
Installing collected packages: pip, wheel
Successfully installed pip-10.0.0 wheel-0.31.0
$ pip

Usage:
  pip <command> [options]
[...]
```

Now if the user has installed another Python 2.7 for, say, python.org or HomeBrew or MacPorts or conda or whatever, the user's shell path may very well have been updated so that typing `python` will invoke that third-party python, not the Apple-supplied one.  And, further, the third-party Python may already have a pip installed that may be able to selfupgrade.  And/or the third-party distributor may have already supplied an updated pip through their normal package update process.  Trying to use the get-pip.py bootstrap may or may not work with the third-party Pythons and may very well screw up their package updates.  So it's probably not a good idea to just tell users to blindly run get-pip.py without knowing which Python they are trying to update.  They might also be trying to update a version of Python 3 (by definition third-party since Apple does not yet ship a version of Python 3).  The get-pip.py instructions wouldn't work in that case either.

And, just for the record, the get-pip bootstrap instructions are not going to work on macOS 10.8 or older because the system `curl` is too old:

```
$ curl https://bootstrap.pypa.io/get-pip.py | python
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
curl: (35) error:1407742E:SSL routines:SSL23_GET_SERVER_HELLO:tlsv1 alert protocol version

```

**Comment by ned-deily** [comment]
Also, for the record, the instructions in https://github.com/pypa/warehouse/issues/3293#issuecomment-378468534 were specifically only for the python.org Python 2.7.14.  Note that the release candidate of 2.7.15 is now available and the final release of 2.7.15 is scheduled for 2018-04-28.  The python.org macOS installers for 2.7.15 (and [rc1](https://www.python.org/downloads/release/python-2715rc1/)) contain both an up-to-date OpenSSL 1.0.2 and pip 9.0.3.

**Comment by brainwane** [comment]
I'm grateful for the details and I accept @ned-deily's judgment here! I suggest I revise this FAQ item so it focuses on getting folks a *diagnosis* but defers the *prescription* to a more interactive support forum. For example (need wordsmithing):

> If you see [one of these error messages], you need to be connecting to PyPI with a newer TLS support library. The specific steps you need to take will depend on your operating system version, where your installation of Python originated (python.org, your OS vendor, or an intermediate distributor), and the installed versions of Python and pip. For help, please go to the #pypa IRC channel, file an issue at pypa/packaging-problems/issues, or post on python-list, including those details and the output of `pip install -vvv upgrade pip`.

**Comment by brainwane** [comment]
@pfmoore may I ask you to double-check the wording in my comment above?

**Comment by pfmoore** [comment]
@brainwane Looks good to me

**Comment by brainwane** [comment]
@alex could you look at this wording to review before I merge?

File: warehouse/templates/pages/help.html [hunk 1]
```diff
@@ -183,6 +185,16 @@ <h3 id="login-problem">{{ login_problem() }}</h3>
...
+        </p>
+        <p><tt>pip install --upgrade -v pip</tt>
+        </p>
+        <p>If you see an error like <tt>There was a problem confirming the ssl certificate</tt> or <tt>tlsv1 alert protocol version</tt> or <tt>TLSV1_ALERT_PROTOCOL_VERSION</tt>, you need to be connecting to PyPI with a newer TLS support library. The specific steps you need to take will depend on your operating system version, where your installation of Python originated (python.org, your OS vendor, or an intermediate distributor), and the installed versions of Python and <tt>pip</tt>. For help, please go to <a href="https://webchat.freenode.net/?channels=%23pypa" target="_blank" rel="noopener">the <tt>#pypa</tt> IRC channel on Freenode</a>, file an issue at <a href="https://github.com/pypa/packaging-problems/issues" target="_blank" rel="noopener">pypa/packaging-problems/issues</a>, or <a href="https://www.python.org/community/lists/" target="_blank" rel="noopener">post to the python-help mailing list</a>, including your OS and installation details and the output of <tt>pip install -vvv upgrade pip</tt>.
```
Comment (alex) [inline warehouse/templates/pages/help.html]: Are we sure the "There was a problem confirming the ssl certificate" error is actually this? I wouldn't expect to see this error message.

**Comment by alex** [comment]
Besides that comment it LGTM.

A follow up might be to mention setuptools; but I think the overwhelming majority of problems have been pip so this is a good MVP.

File: warehouse/templates/pages/help.html (hunk 1 above)
Comment (brainwane) [reply warehouse/templates/pages/help.html]: Some users do get "There was a problem confirming the ssl certificate" as part of the error for this issue -- I looked through https://github.com/issues?utf8=%E2%9C%93&q=%22there+was+a+problem+confirming+the+ssl+certificate%22+user%3Apypa+ and verified.

Comment (alex) [reply warehouse/templates/pages/help.html]: "There was a problem confirming the ssl certificate: [SSL: TLSV1_ALERT_PROTOCOL_VERSION] tlsv1 alert protocol version" gosh that's an awful error message :-(

Okey doke! Thanks for putting this together.

**Comment by brainwane** [comment]
Thank you @alex. Merged.

@ncoghlan this is at least the start of some better docs for users on this.

**Pull request merged**